### PR TITLE
Xlink functionality for the dash-live-source-simulator

### PR DESCRIPTION
**Pull request by AdithyanIlangovan**
Currently, in the dash-live-source- simulator multiple periods are generated by specifying periods_n in the url. We would like to make few of these multi periods to be accessible via xlink.

This functionality has been implemented in the simulator. The url should look something like:
>http://server/periods_n/xlink_m/Manifest.mpd

The above url would result in creating a multi-period live service with n periods per hour, where m period per hour would be accessed via xlink. One xlink period is created for every n/m periods.

### CODE REVIEW

**Comment by TobbeMobiTV** [comment]
Just tried to run the tests ./run_tests.sh and the new test fails. 

     ERROR: testMpdPeriodReplaced (dashlivesim.tests.test_xlinkperiod.TestXlinkPeriod)
     Check whether appropriate periods have been replaced by in .mpd file
     Traceback (most recent call last):
     ...
      xlink_period = '<Period xlink:href="%s%s+%s.period" xlink:actuate="onLoad" xmlns:xlink="http://www.w3.org/1999/xlink"></Period>' %(base_url[0],filename,period_id)
     IndexError: list index out of range


Furthermore, pylint is complaining about bad whitespace, too long lines. Run 

      pylint dashlib

in the dashlivesim.

There are also misspellings in comments, so please clean up the code, and I'll have another look.

**Comment by AdithyanIlangovan** [comment]
Strange. I just tried running the test in my system and it works.

test

Just to be sure, I downloaded the .zip file and ran the tests from another system. And it worked.

I will commit a new version with cleaned up comments and the pylint taken care of.

**Comment by TobbeMobiTV** [comment]
The error occurred when the state `mpdprocessor.SET_BASEURL = False` which depended on the order of the tests. I've updated other tests (commit 282fd65) so that the order should no longer play any role. However, I recommend that you set `mpdprocessor.SET_BASEURL = True` in your test similar to  what is done in `test_multi_baserurl.py`.

**Comment by AdithyanIlangovan** [comment]
Thank you for the comment. I will do the changes.
So, does this mean the test/code is working correctly for you now?
I will have to just clean up the spaces and comments now, right?

**Comment by TobbeMobiTV** [comment]
Yes, your new test worked as I turned on the SET_BASEURL attribute, and recommend that you do that change.

**Comment by AdithyanIlangovan** [comment]
I have cleaned up the code now.

**Comment by TobbeMobiTV** [comment]
Pylint is OK now, but the testMpdPeriodReplaced is useless as far as I can tell.
The period_id_all value is an empty array. That is because there are 3 periods in the output that all start with `<Period xlink...` but none that start with `<Period id=...` that you're looking for.

I also wonder if you don't need a way to specify a completely independent URL for the xlink.
Can you give a full example of how this is intended to be used and fix the test?

**Comment by AdithyanIlangovan** [comment]
Thanks for the comments.
>if you don't need a way to specify a completely independent URL for the xlink.

No, this functionality is only meant to replace few of the existing mutli-periods with xlinks. 
Please refer to Section 2 of this document for further details:
[xlink.pdf](https://github.com/Dash-Industry-Forum/dash-live-source-simulator/files/92134/xlink.pdf)
Inputting independent URL for the xlink could be a future feature. 

And regarding the test. For the current set of values:
>nr_period_per_hour = 2
>nr_xlink_periods_per_hour = 2

If the implementation works correctly, all the periods in the output will be replaced with xlink and that is happening right now. 
If some of the periods were not replaced, some will have the <Period id=  element and then the final check will fail. 

A better test would be is : I repeat the same test for a combination of values such as :
>nr_period_per_hour = 10 and nr_xlink_periods_per_hour = 2
>nr_period_per_hour = 10 and nr_xlink_periods_per_hour = 5

And see it if is working for all. What do you think?

**Comment by waqarz** [comment]
As far as I can see, the test is validating to an extent, if this is the ideal test I cannot say, but it is useful to its scope.

The requirement this feature implementation fulfills is to add xlink with a configurable frequency as documented in the attachment before, which already is a good test for client if it is able to resolve xlink correctly or not. This functionality can be extended later by adding different content.

File: dashlivesim/tests/test_xlinkperiod.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    def testMpdPeriodReplaced(self):
+        " Check whether appropriate periods have been replaced by in .mpd file"
+        nr_period_per_hour = 2
+        nr_xlink_periods_per_hour = 2
+        urlParts = ['livesim', 'periods_%s' %nr_period_per_hour, 'xlink_%s' %nr_xlink_periods_per_hour, 'testpic_2s', 'Manifest.mpd']
+        dp = dash_proxy.DashProvider("10.4.247.98", urlParts, None, VOD_CONFIG_DIR, CONTENT_ROOT, now=10000)
+        d = dp.handle_request()
+        period_id_all = findall('Period id="([^"]*)"', d)
+        # Find all period ids in the .mpd file returned.
```
Comment (TobbeMobiTV) [inline dashlivesim/tests/test_xlinkperiod.py]: period_id_all = [].
Isn't it enough to check that there are no elements in this array?

**Comment by TobbeMobiTV** [comment]
@AdithyanIlangovan Thanks for the link. I had missed that before. Will read the document, and then I'll probably be enlighted. Please change the check so that not all periods are xlink.

**Comment by TobbeMobiTV** [comment]
I read the xlink.pdf document, and think I understand what you want to do. The document should probably be added to this site once we include this. Maybe as a Wiki-page. For this to be testable by a player, one still need to put the xlink data on the same server or introduce some HTTP redirect method. Do you have a server/player combination where you can test this?

**Comment by AdithyanIlangovan** [comment]
>Do you have a server/player combination where you can test this?
In our local server here, we have the dash.js player. When I run the current dash-live server/player, the content is being played as it should.

**Comment by TobbeMobiTV** [comment]
@AdithyanIlangovan Good that you can test it locally. Btw, the test that you do is relying on the order of the attributes of the `<Period>` entity. A more robust test would use an XML parser and check the attributes by name rather than doing a string comparison. Probably not an issue, but I'm running into the same situation in another context so I have it on my mind.

**Comment by AdithyanIlangovan** [comment]
I see what you mean. In future, I would parse and extract the "id" attribute and use that for the test.

**Comment by TobbeMobiTV** [comment]
This version is now running at http://vm2.dashif.org/livsim-dev/

**Pull request merged**